prov/util/src/util_cq.c

```diff
@@ -256,53 +238,38 @@ ssize_t ofi_cq_readfrom(struct fid_cq *cq_fid, void *buf, size_t count,
 	if (count > ofi_cirque_usedcnt(cq->cirq))
 		count = ofi_cirque_usedcnt(cq->cirq);
 
-	for (i = 0; i < (ssize_t)count; i++) {
+	for (i = 0; i < (ssize_t) count; i++) {
 		entry = ofi_cirque_head(cq->cirq);
-		if (OFI_UNLIKELY(entry->flags & (UTIL_FLAG_ERROR |
-						 UTIL_FLAG_OVERFLOW))) {
-			if (entry->flags & UTIL_FLAG_ERROR) {
-				struct util_cq_aux_entry *aux_entry =
-						container_of(cq->aux_queue.head,
-							     struct util_cq_aux_entry,
-							     list_entry);
-				if (aux_entry->comp.err) {
-					/* This handles case when the head of aux_queue is
-					 * an error entry.
-					 *
-					 * NOTE: if this isn't an error entry, we have to handle
-					 * overflow entries and then the error entries to ensure
-					 * ordering. */
-					if (!i)
-						i = -FI_EAVAIL;
-					break;
-				}
+		if (!(entry->flags & UTIL_FLAG_AUX)) {
+			if (src_addr && cq->src)
+				src_addr[i] = cq->src[ofi_cirque_rindex(cq->cirq)];
+			cq->read_entry(&buf, entry);
+			ofi_cirque_discard(cq->cirq);
+		} else {
+			assert(!slist_empty(&cq->aux_queue));
+			aux_entry = container_of(cq->aux_queue.head,
+						 struct util_cq_aux_entry,
+						 list_entry);
+			assert(aux_entry->cq_slot == entry);
+			if (aux_entry->comp.err) {
+				if (!i)
+					i = -FI_EAVAIL;
+				break;
 			}
-			if (entry->flags & UTIL_FLAG_OVERFLOW) {
-				assert(!slist_empty(&cq->aux_queue));
+
+			if (src_addr && cq->src)
+				src_addr[i] = aux_entry->src;
+			cq->read_entry(&buf, &aux_entry->comp);
+			slist_remove_head(&cq->aux_queue);
+
+			if (!slist_empty(&cq->aux_queue)) {
```

Comment:
If the aux queue is empty, don't you want to do an ofi_cirque_discard?

Reply:
Yes, I believe you're right.  discard is needed in that case.  Thanks for catching this.  I'll update.